### PR TITLE
th/checkpoint

### PR DESCRIPTION
**Pull request by thom311**
Some fixes, refactoring, and new features for checkpoints

### CODE REVIEW

File: src/nm-dbus-utils.c [hunk 1]
```diff
@@ -147,6 +147,36 @@ nm_dbus_utils_g_value_set_object_path_array (GValue *value,
 	g_value_take_boxed (value, paths);
 }
 
-/*****************************************************************************/
+void
+nm_dbus_utils_g_value_set_object_path_from_hash (GValue *value,
+                                                 GHashTable *hash /* is has keys of NMDBusObject type. */,
```
Comment (bengal) [inline src/nm-dbus-utils.c]:
```
/* is has keys of NMDBusObject type. */,
   ^^^^^

File: introspection/org.freedesktop.NetworkManager.xml [hunk 2]
```diff
@@ -251,6 +251,27 @@
       <arg name="result" type="a{su}" direction="out" />
     </method>
 
+    <!--
+        CheckpointAdjustRollbackTimeout:
+        @add_timeout: number of seconds from ~now~ in which the
+          timeout will expire. Set to 0 to disable the timeout.
+          Note that the added seconds start counting from now,
+          not Created timestamp or the previous expiriation
```
Comment (bengal) [inline introspection/org.freedesktop.NetworkManager.xml]: expiration

File: libnm-core/nm-dbus-interface.h [hunk 3]
```diff
@@ -842,6 +842,18 @@ typedef enum {
  *   delete any new connection added after the checkpoint (Since: 1.6)
  * @NM_CHECKPOINT_CREATE_FLAG_DISCONNECT_NEW_DEVICES: upon rollback,
  *   disconnect any new device appeared after the checkpoint (Since: 1.6)
+ * @NM_CHECKPOINT_CREATE_FLAG_ALLOW_OVERLAPPING: by default, creating
+ *   a checkpoint fails if there are already existing checkoints that
+ *   reference the same devices. With this flag, creation of such
+ *   checkpoints is allowed, however, if an older checkpoint
+ *   that references overlapping devices gets rolled back, it will
+ *   automatically destroy this checkpoint during rollback. This
+ *   allows to create several overlappig checkpoints in parallel,
```
Comment (bengal) [inline libnm-core/nm-dbus-interface.h]: overlapping

**Comment by bengal** [comment]
> checkpoint: allow resetting the rollback timeout via D-Bus

> The libnm API only implements the async form (at least for now).
> Sync methods are fundamentally wrong with D-Bus, and it's probably
> not needed. Also, follow glib convenction, where the async form
> doesn't have the _async name suffix.

If so, can you also rename the existing checkpoint functions to follow
the same pattern? They are not released yet and so it's not an API
breakage.

**Comment by thom311** [comment]
fixed, repushed.

Note the FIXME comments added by the last commit. There is still follow-up work to do (maybe not as part of this branch).

**Comment by bengal** [comment]
LGTM

**Comment by thom311** [comment]
Thanks. merged

**Pull request closed without merging**